### ISSUE TITLE
VR mode shows black screen both on Oculus quest and Cardboard

### ISSUE THREAD

**Issue by eggnot** [magcius/noclip.website#490]
Both Oculus and smartphone showing black screen entering VR. There is no option "XR sandbox..." on current chrome version on both as well

**Comment by magcius** [comment]
Yeah, the VR support is currently broken. Unfortunately, I do not have a headset or a way to test it myself. Last I checked, it worked fine on the XR emulation extension.

**Comment by jrupinski** [comment]
Tested on Oculus Quest 2 64 GB model, both on a local server and on the [noclip.website](noclip.website).
The VR Option is there and it works in compatible browsers (both _Firefox Reality_ and built-in _Oculus browser_) - but on every tested browser the scenery seems to be broken while rendering on certain titles, both while using VR and non-VR view of the website, so it's a per-title problem. 

Doesn't work properly:
- Team Fortress 2 (and other source titles) - only renders water (tested on 2-fort map)
- Mario Kart Wii - renders all-black textures, aside from speed-up tracks (tested on Luigi's circuit map)

What works:
- Super Mario 64 DS renders properly
- Psychonauts both renders flawlessly, and controller move speed seems to be optimized for quest 2 perfectly
- Luigi's Mansion for the Gamecube renders fine

**Comment by magcius** [comment]
Unless someone is able to debug the actual device, I can't really contribute here. I am happy to assist someone who wants to understand how the VR mode is architected, but given that it works on the WebXR emulation extension, I do not think I am doing anything majorly wrong here...

**Comment by magcius** [comment]
Thanks to some community members, I identified one potential issue fixed in 8c36ec1ffa81c61dc56bf7df9a3c1dbb4cc5654d.

**Comment by jrupinski** [comment]
This seems to fix the issue with VR mode on PC based on community reports.

Standalone VR headsets still seem to have rendering issues in some specific games.
Fixing this would probably require going through each of the games to find the problematic ones, and comparing (maybe replacing?) code with one that renders correctly (like Super Mario 64 DS). 

Resources for someone with a Quest 2 headset and willing to try and fix this:
- Viewer::renderWebXR is the entry point
- https://developer.oculus.com/documentation/web/webxr-perf-renderdoc/

A video to illustrate the problem:

https://user-images.githubusercontent.com/9074800/162080727-99ef1324-cb22-4c5d-b485-ca903ad5b193.mp4

**Comment by chandler05** [comment]
I've got an HTC Vive Index and I've noticed some of the maps (Both Mario Galaxies, Wii Sports, Wind Waker, and more) render strangely to where two very different images render in both eyes and cause the maps to not show up properly.

Here's some videos to show:

Funcitoning Map (Okami):
https://github.com/magcius/noclip.website/assets/66492208/387b40e6-aa59-40a8-a4bc-e424a694b5fa

Broken Map (Super Mario Galaxy):
https://github.com/magcius/noclip.website/assets/66492208/09af9ab8-621b-474b-ad85-1308d809b8a2

**Comment by magcius** [comment]
I tried using the [WebXR API Emulator](https://chrome.google.com/webstore/detail/webxr-api-emulator/mjddjgeghkdijejnciaefnkjmkafnnje) extension, and the render looked correct to me... I do not have a personal device to test with.

**Comment by maximxlss** [comment]
My take:
- Device: Quest 2
- Map: Luigi Circuit - Mario Kart Wii
- The app doesn't produce any relevant errors in the console.
- Screenshot: <img width="1381" alt="image" src="https://github.com/magcius/noclip.website/assets/29152154/6441dc95-cd23-4943-8822-6a5d3ddfe82d">
- Screenshots of [webglreport.com](https://webglreport.com) on the device as well as on my phone (both Qualcomm Adreno): [https://imgur.com/a/0OZYhJt](https://imgur.com/a/0OZYhJt) (KHR_parallel_shader_compile and OES_draw_buffers_indexed extensions are missing on the device)

I have no experience working with low-level graphics, so I don't think I can figure it out, but I'll continue my research.

**Comment by maximxlss** [comment]
What doesn't work (black textures):
- All Wii games
- Some Gamecube games (e.g. Metroid Prime)
What does work:
- Some Gamecube games (e.g. Luigi's Mansion)
- All other games I tried.
